### PR TITLE
Don't substitute into the targeted instructions of an associated constant.

### PR DESCRIPTION
**Pull request by zygoloid**
When an instruction makes an absolute reference to another instruction, such as when `assoc_const` refers to the declaration of the associated constant in an interface, substitution into that instruction should not substitute into the referenced instruction.

Mark the corresponding `InstId` fields in the typed instructions as being absolute by giving them a distinct ID type that `Subst` doesn't substitute into. This formation of unnecessarily complicated SemIR that could in some cases lead to a CHECK failure when printing formatted SemIR because the same instruction ends up in multiple scopes.

### CODE REVIEW

**Comment by zygoloid** [comment]
I've split this PR into three commits: the first adds the test and shows the old behavior (with the CHECK disabled), the second fixes the bug, and the third updates the test to show new new behavior (and re-enables the CHECK). This should hopefully give a better idea of what this PR is fixing.

**Comment by zygoloid** [comment]
#4336 is blocked on this after merge because the check added in #4333 (landed in #4221) starts firing on one of the #4336 tests due to this bug.

File: toolchain/sem_ir/ids.h
Comment (geoffromer) [inline toolchain/sem_ir/ids.h]: The phrasing of this comment makes it sound kind of like "absoluteness" is a property of the inst itself, but based on the code, I gather it's actually a property of the ID. In other words, in principle you could have an `AbsoluteInstId` and an `InstId` that refer to the same inst. Is that right?

File: toolchain/sem_ir/ids.h
Comment (geoffromer) [inline toolchain/sem_ir/ids.h:98]: This means we inherit the base class `Print`, but might we want the difference to be visible in the printed output?

File: toolchain/sem_ir/ids.h
Comment (geoffromer) [inline toolchain/sem_ir/ids.h:103]: The implicit conversion makes me a little nervous, especially since the conversion in the other direction is implicit as well. What would we lose if it was explicit?

File: toolchain/sem_ir/ids.h
Comment (zygoloid) [reply toolchain/sem_ir/ids.h:98]: I don't think we do; we're trying to attach an extra bit of information onto the typed inst field, not change how the field behaves.

File: toolchain/sem_ir/ids.h
Comment (zygoloid) [reply toolchain/sem_ir/ids.h:103]: It'd just introduce noise into all the places that construct typed insts, explicitly performing a conversion they don't care about.

File: toolchain/sem_ir/ids.h
Comment (zygoloid) [reply toolchain/sem_ir/ids.h]: Yes, that's right. I've updated the comment to try to make this clearer.

**Review by geoffromer — APPROVED** [review]

**Pull request merged**
